include/bluetooth/bluetooth.h

```diff
@@ -71,6 +79,56 @@ int bt_enable(bt_ready_cb_t cb);
  */
 int bt_set_id_addr(const bt_addr_le_t *addr);
 
+/** @brief Get the currently configured identities.
+ *
+ *  Returns an array of the currently configured identity addresses. To
+ *  make sure all available identities can be retreived, the number of
+ *  elements in the @a addrs array should be CONFIG_BT_ID_MAX.
+ *
+ *  @param addrs Array where to store the configured identities.
+ *  @param count Should be initialized to the array size. Once the function
+ *               returns it will contain the number of returned identies.
+ */
+void bt_id_get(bt_addr_le_t *addrs, size_t *count);
+
+/** @brief Create a new identity.
+ *
+ *  Create a new identity using the given address and IRK. This function
```

Comment:
Just looking at the API docs, I'd suggest clarifying the difference between calling this before and after bt_enable(). Maybe they should even be two separate functions?

Reply:
Separate functions seems a bit excessive (I have also no idea how to name them intuitively in that case). I do have at least some documentation on the differences in usage, I wonder if you were looking at an earlier revision of the patch set?

Reply:
> "can be used to override the controller's public address" does this mean I have to do more to make this happen, or will it always happen.

It will always happen. It would get overwritten though if there was something stored in flash and the app calls `settings_load()`. Btw, the design is like this to keep the existing `bt_set_id_addr()` functionality which was added for Zephyr 1.12. Now that API is just a wrapper for `bt_id_create()`.

> "(in case it has one)" what happens if it doesn't?

If the controller has no public address, then `bt_id_create()` can be used before bt_enable() to populate pre-defined identity addresses, starting with the default identity. The stack will auto-generate a static random address for the default identity if one is not found in flash and was also not provided through bt_id_create().

> Is the id returned from a call before bt_enable() equivalent to ids returned after bt_enable() until next reset? I.e. can it be used for all the same things?

Yes, the id is the same, it's just an array index to the stack-internal array of identities. Only difference is that before bt_enable() nothing will get written to flash, i.e. if an app does this, then the app needs to repeat the same `bt_id_create()` call upon every bootup.

> Are _ANY/NULL parameters valid before bt_enable()?

Yes. No difference there.

> Does the controller's default public address have an ID?

Yes, it will be BT_ID_DEFAULT (i.e. 0) if available and not overridden by calling `bt_id_create()` before `bt_enable()`.

Reply:
Can the identity address be changed with bt_set_id_addr()? With the limit of 10 identies we will run out of identites quite fast if we do not recycle old ones. I.e in the mouse use case where you have a mouse that can support 3 identies. What happens when we want to add a new bond? We can either create a new identity with a new address (will reach 10 fast) or can we change the address for an existing identity.

Reply:
I'd then consider adding something like the following:
```
int bt_id_reset(u8_t id, bt_addr_t *addr, u8_t *irk);
```

This would reinitialize the identity and remove any bondings it had.

The alternative approach would be to modify `bt_id_create()`:
```
#define BT_ID_ANY 0xff

int bt_id_create(u8_t id, bt_addr_t *addr, u8_t *irk);
```

If `id` is anything else than BT_ID_ANY it would need to be an already created id, otherwise the function would return failure. When given an already created id this would behave the same way as my `bt_id_reset()` proposal.

Reply:
So `bt_id_create()` will override ids in some instances but not others? What happens if persistent identities are created in between non-persistent ids? Do they get consecutive ids?

Reply:
It will override if you give it a valid id. I guess BT_ID_NEW would be more intuitive than BT_ID_ANY in my example.

Mixing persistent (stack-generated address) and non-persistent (app-generated address) IDs is something I didn't think about. The API as drafted currently would allow it, but I fail to imagine a good use case for it. Do you have one in mind? The way the addresses and IRKs are stored in flash is using the stack-internal array as-is. I.e. with the current implementation any persistent id would cause all ids to be written to flash.

Reply:
>Do you have one in mind?

Not really, but users will use it in all sorts of ways, so it might be good to document it.

Well maybe there is one:

 1. Call bt_id_create() to override controller's address
 2. bt_enable()
 3. settings_load()
 4. Call bt_id_create() to generate random ids.

Will 4. cause 1. to be written to flash?

Reply:
How about adding a

    bt_id_store()

Which stores to flash whatever would be returned by `bt_id_get()` at that time.

Reply:
There are a few issues I see with that:

 - We don't have explicit "store" APIs for anything else in the stack. Information (pairing keys, GATT CCC, etc) is stored implicitly
 - You risk getting into a situation where the app doesn't call `bt_id_store()` but the stack goes and performs a pairing and ends up storing information related to the id, but not the id itself.

Reply:
I'd like to add that currently the storing of ids is deferred through the system work queue, i.e. the app can do multiple bt_id_create(); bt_id_create(); ... calls but it should still just result in just a single write to flash.

Reply:
> We don't have explicit "store" APIs for anything else in the stack. Information (pairing keys, GATT CCC, etc) is stored implicitly

I see

> You risk getting into a situation where the app doesn't call bt_id_store() but the stack goes and performs a pairing and ends up storing information related to the id, but not the id itself.

Can't this situation already happen? Not all ids are stored persistently, and the stack can't really know which is which.

Reply:
> Can't this situation already happen? Not all ids are stored persistently, and the stack can't really know which is which.

What the stack assumes is that if it is given pre-defined values, then the app is responsible for providing these always. They might be hard-coded into the app (this is the use case why `bt_set_id_addr()` was added), come from some factory-programmed register, special flash location or something else. So it would be an application bug if it fails to keep providing these ids for every power cycle. I can try to make this clear in the documentation.

Reply:
> What the stack assumes is that if it is given pre-defined values, then the app is responsible for providing these always.

I see. This gets a bit more complicated when considering DFU updates, though. IMO the user should have the possibility to store even predefined identities if that is more convenient for them.

Reply:
Would it be an option then to always store the identities post-bt_enable, regardless of being stack-generated or pre-defined? Before bt_enable I wouldn't touch flash however, since at that point we don't know the flash content and would possibly be (over)writing blindly to flash.

Reply:
As a user, I think I would find that less confusing. Maybe a Kconfig could contol it? I don't know the conventions.